pkg/guardian/prisoner.go

```diff
@@ -0,0 +1,115 @@
+package guardian
+
+import (
+	"fmt"
+	lru "github.com/hashicorp/golang-lru"
+	"net"
+	"sync"
+	"time"
+)
+
+type Prisoner struct {
+	IP     net.IP    `yaml:"ip" json:"ip"`
+	Jail   Jail      `yaml:"jail" json:"jail"`
+	Expiry time.Time `yaml:"expiry" json:"expiry"`
+}
+
+type prisonersCache struct {
+	// cache is a lru cache safe for concurrent use by multiple go routines.
+	cache *lru.Cache
+
+	// we need a mutex in order to serialize the execution of multiple actions to the cache (e.g. get all prisoners)
+	mutex sync.RWMutex
```

Comment:
After reading the LRU docs, I don't think this is necessary. `prisonersCache` is a thin wrapper around `lru.Cache` which is threadsafe.

Reply:
There are 2 reasons why I decided to go with this thin wrapper.
1. **Better enforcement of type safety**. Since the LRU cache utilizes `interface{}` as an argument for several function calls, I'm a bit worried about exposing the lru cache directly. By providing a thin wrapper, we can easily tell whether there are any issues w.r.t inserting incorrect types into the cache without searching other parts of the code base. However, I guess it would require me to place this in a separate package to truly ensure that other parts of the codebase do not access the lru cache directly. 
2. **While the cache is thread safe, there are race conditions when doing multiple operations on the cache.** For example, see the `getPrisoners` method. The LRU cache exposes a `Keys()` function, but no means to read the entire cache and obtain all of the values. So in order to do this, we have to iterate through the keys and get each value. In order to make sure other goroutines aren't updating/removing/adding keys, some concurrency mechanism needs to be in place. I chose a RWLock in this case.

Reply:
I'm not against the wrapper, I think the additional mutex is unnecessary though.

It looks like the [implementation](https://github.com/hashicorp/golang-lru/blob/v0.5.4/lru.go#L137) is simple enough that you could just fork it and add a `func (c *Cache) Snapshot() map[interface{}]interface{}` method that does an atomic snapshot using the internal LRU mutex.

Reply:
I don't consider this to be a merge blocker, I'm bringing it up as a possible scalability concern.

Reply:
also consider https://redis.io/topics/lru-cache

Reply:
> I'm not against the wrapper, I think the additional mutex is unnecessary though. It looks like the implementation is simple enough that you could just fork it and add a func (c *Cache) Snapshot() map[interface{}]interface{} method that does an atomic snapshot using the internal LRU mutex.

Agreed. I will fork it. Maybe they will accept this change upstream too. 

> also consider https://redis.io/topics/lru-cache

I was looking into this, but unfortunately I don't think any of the supported eviction policies would work in this use case. 

For example, `allkeys-lru` applies the eviction policy to all keys in Redis, so we could potentially evict keys that store guardian configuration values that are required. 

The best fits would probably be `volatile-lru` or `volatile-ttl`, but even then, we could be evicting counters that we depend on for rate limiting.

Reply:
Ah, just thought of the reason why `Snapshot` isn't used upstream. It's because it basically invalidates the entire cache by no longer keeping track of the last, legitimate least recent usage. 

Anyways, I think we can live without this function. I will remove it since it's just used by `guardian-cli` and we don't really need a cache in that case.